### PR TITLE
Updated up to the current standards + DateTime family classes support

### PR DESCRIPTION
**Pull request by nilopc**
Please take some time to review this. I'm open for discussion any time

Nothing has been broken, but there's a few changes here and there. :)

Oh and thanks for open-sourcing this!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,9 +2,15 @@ language: php
 php:
   - "5.4"
   - "5.5"
+  - "5.6"
+  - "nightly"
```
Comment (cjsaylor) [inline .travis.yml]: @nilopc This should be in the allow_failure matrix.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
 # Json Serializer for PHP
 
+##### Forked
```
Comment (cjsaylor) [inline README.md]: @nilopc This shouldn't be included in the PR for merge.

**Comment by cjsaylor** [comment]
@nilopc It would be preferable to have all the psr4 based moved in a separate PR and contain just the actual behavior changes to this PR.

File: .travis.yml [hunk 3]
```diff
@@ -2,9 +2,15 @@ language: php
...
   - hhvm
 
 before_script:
   - composer install --dev
 
 script: ./vendor/bin/phpunit --coverage-text
+
+matrix:
+     allow_failures:
+         - php: hhvm
```
Comment (jrbasso) [inline .travis.yml]: @nilopc hhvm was working fine before the PR, I would like to keep it compatible. Do you know why those tests are failing?

File: .travis.yml (hunk 1 above)
Comment (jrbasso) [inline .travis.yml]: I actually would use 7.0. It is the same thing, but when the 7.0 is released we don't need to update the travis file. Also, 7.0 is already in beta, it is stable enough to keep as a required environment to test IMO.

File: README.md [hunk 4]
```diff
@@ -8,19 +13,20 @@ This is a library to serialize PHP variables in JSON format. It is similar of th
 but the output is a string JSON encoded. You can also unserialize the JSON generated by this tool and have you
 PHP content back.
 
-Supported features:
+**Supported features**:
 - Encode/Decode of scalar, null, array
-- Encode/Decode of objects
+- Encode/Decode of objects, including all the [`DateTime family objects`](http://php.net/manual/en/book.datetime.php)  (except `DatePeriod`).
```
Comment (jrbasso) [inline README.md]: @nilopc Any reason why not to support `DatePeriod`?

**Comment by jrbasso** [comment]
I agree with @cjsaylor in separate the PSR-4 changes from this PR to another so the changes are clear.

Also, when you separate the PR to psr-4 I would recommend to make one commit to rename the files + change the paths on `composer.json` and another commit to change the spacing. Otherwise git understand as different files and makes a delete/add files instead of rename and the file history goes way.

**Comment by jrbasso** [comment]
By the way, changes look good, :clap: for contributing.

File: .travis.yml (hunk 1 above)
Comment (cjsaylor) [inline .travis.yml]: :+1: , but still needs a 7.0 update in this PR.

File: .travis.yml (hunk 3 above)
Comment (nilopc) [inline .travis.yml]: I'm thinking of a way of solving this at the moment... for some reason it's just DateTimeZone and DateInterval classes that are failing.

File: README.md (hunk 4 above)
Comment (nilopc) [inline README.md]: I tried serializing it and after checking out the results i found out it needs a very specific and concrete serialization strategy on its own.

File: README.md (hunk 2 above)
Comment (nilopc) [inline README.md]: Will remove this :+1:

**Comment by nilopc** [comment]
@jrbasso @cjsaylor Added hack-specific code to make the class pass in all cases: 

- https://github.com/nilportugues/json-serializer/commit/2459a6c55e633137ef95a6087aae5d480cedd1d8

- https://travis-ci.org/nilportugues/json-serializer/builds/69459189 :+1: 

This will need refactoring though because the class is getting huge. 

Maybe splitting it in two or three files would make sense. How do you feel about this?

**Comment by jrbasso** [comment]
@nilopc The hack seems to be ok.

I would split the class just if that makes sense. I wouldn't split just because it is too long. I prefer to keep things logically together.
Do you have any proposal for the split?
